ui/src/utils/formatters.ts

```diff
@@ -58,15 +58,46 @@ export const compareButIgnoreLeadingDashes = (a: string, b: string) => {
 	return aClean.localeCompare(bClean);
 };
 
+interface IGetVcsUrlUniversalParams {
+	serviceInfo: IServiceMapValue | undefined;
+	repo: string;
+	filename: string | undefined;
+	branch: string | undefined;
+	commit: string | undefined;
+	line_no: string | undefined;
+}
+
+interface IGetVcsUrlServiceParams {
+	commitBase: string; // The string that goes between repo and commit in the url
+	lineParam: string; // The string that goes before the line number in the url
+}
+
+const nonFileLocations = new Set(["Commit Message"]);
+
+const getVcsUrl = (
+	universal: IGetVcsUrlUniversalParams,
+	service: IGetVcsUrlServiceParams,
+) => {
+	const base = `${universal.serviceInfo?.url ?? ""}/${universal.repo}`;
+	const refPath = universal.commit || universal.branch;
+	const lineText =
+		universal?.line_no !== "0"
+			? `#${service.lineParam}${universal.line_no}`
+			: "";
+
+	if (!universal.filename) {
+		return base;
+	} else if (nonFileLocations.has(universal.filename)) {
+		// Link to the commit directly if filename is special
+		return `${base}/${service.commitBase}/${refPath}`;
+	} else {
+		return `${base}/${service.commitBase}/${refPath}/${universal.filename}${lineText}`;
+	}
+};
+
 export const vcsHotLink = (row: RowDef) => {
 	let url = "";
-	if (
-		row &&
-		row.service &&
-		row.service.length > 0 &&
-		row.repo &&
-		row.repo.length > 0
-	) {
+	if (row?.service?.length > 0 && row?.repo?.length > 0) {
```

Comment:
Simplified this a bit with optional chaining. Not tied to this, if you'd prefer if we limit unnecessary changes